app/Community/Actions/SendDailyDigestAction.php

```diff
@@ -0,0 +1,119 @@
+        foreach ($ids as $type => $typeIds) {
+            $titles[$type] = match ($type) {
+                SubscriptionSubjectType::ForumTopic->value => ForumTopic::whereIn('id', $typeIds)->pluck('title', 'id'),
+                default => [],
+            };
+        }
+
+        // build the data to pass to the mail script
+        $notificationItems = [];
+        foreach ($delayedSubscriptions as $delayedSubscription) {
+            // if all the new posts have been deleted or aren't visible, ignore it
+            $handler = $this->getHandler($delayedSubscription->subject_type);
+            $count = $handler->getUpdatesSince($delayedSubscription);
+            if ($count > 0) {
+                $notificationItems[] = [
+                    'type' => $delayedSubscription->subject_type->value,
+                    'title' => $titles[$delayedSubscription->subject_type->value][$delayedSubscription->subject_id] ?? '(untitled)',
+                    'link' => $handler->getLink($delayedSubscription->subject_id, $delayedSubscription->first_update_id),
+                    'count' => $count,
+                ];
+            }
+        }
+
+        // send the mail
+        Mail::to($user->EmailAddress)->queue(
+            new DailyDigestMail($user, $notificationItems)
+        );
+    }
+
+    private function getHandler(SubscriptionSubjectType $subjectType): BaseDelayedSubscriptionHandler
+    {
+        return match ($subjectType) {
+            SubscriptionSubjectType::ForumTopic => new ForumTopicDelayedSubscriptionHandler(),
+
+            // this is wrong, but we have to return something to satisfy phpstan.
+            // other cases will be filled in as the calls are updated.
+            default => new ForumTopicDelayedSubscriptionHandler(),
+        };
+    }
+}
+
+abstract class BaseDelayedSubscriptionHandler
+{
+    /**
+     * Gets the number of updates that have been made by users other than the subscriber since the notification was delayed.
+     */
+    abstract public function getUpdatesSince(UserDelayedSubscription $delayedSubscription): int;
+
+    /**
+     * Gets a link to the first updated subrecord of the subject.
+     */
+    abstract public function getLink(int $subjectId, int $firstUpdateId): string;
+}
+
+class ForumTopicDelayedSubscriptionHandler extends BaseDelayedSubscriptionHandler
+{
+    public function getUpdatesSince(UserDelayedSubscription $delayedSubscription): int
+    {
+        return ForumTopicComment::query()
+            ->where('forum_topic_id', $delayedSubscription->subject_id)
+            ->where('id', '>=', $delayedSubscription->first_update_id)
+            ->where('author_id', '!=', $delayedSubscription->user_id)
+            ->where('is_authorized', true)
+            ->count();
+    }
+
+    public function getLink(int $subjectId, int $firstUpdateId): string
+    {
+        return route('forum-topic.show', ['topic' => $subjectId, 'comment' => $firstUpdateId]) . '#' . $firstUpdateId;
+    }
+}
```

Comment:
nit: Unless `BaseDelayedSubscriptionHandler` and `ForumTopicDelayedSubscriptionHandler` are going to stay private to the action forever, we should probably break them both out to their own individual separate files so we're compliant with [the PSR-4 autoloading standard](https://www.php-fig.org/psr/psr-4/).

Right now, both of these classes ignore `namespace App\Community\Actions;`, effectively making them globally-namespaced.